### PR TITLE
Delete consent receipts 

### PR DESCRIPTION
**Pull request by VivekVinushanth**
### Proposed changes in this pull request
When application is being disabled, consents provided against those applications should be revoked as well. 

**Partially Fixes**
https://github.com/wso2/product-is/issues/20378

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java [hunk 1]
```diff
@@ -135,6 +136,50 @@ public boolean doPostDeleteApplication(String applicationName, String tenantDoma
...
+     * When an application is disabled, it will delete all relevant receipts issued against that application.
+     *
+     * @param serviceProvider Service provider which is disabled.
+     * @param tenantDomain    Tenant domain of the application.
+     * @param userName        Username of the person who does the deletion.
+     * @return true.
+     * @throws IdentityApplicationManagementException IdentityApplicationManagementException.
+     */
+    @Override
+    public boolean doPostUpdateApplication(ServiceProvider serviceProvider, String tenantDomain, String userName)
```
Comment (janakamarasena) [inline components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java:149]: seems like there is lot of common things in this method and `doPostDeleteApplication` lets extract this into a common method and use in both places.

File: components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java [hunk 2]
```diff
@@ -135,6 +136,50 @@ public boolean doPostDeleteApplication(String applicationName, String tenantDoma
...
+            return true;
+        }
+        String applicationName = serviceProvider.getApplicationName();
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Deleting consents since application: %s, in tenant domain: %s is disabled.",
+                    applicationName, tenantDomain));
+        }
+        try {
+            List<ReceiptListResponse> receiptListResponses = consentManager.searchReceipts(consentSearchLimit, 0,
+                    "*", tenantDomain, applicationName, null, null);
```
Comment (janakamarasena) [inline components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java:163]: magic numbers... lets move 0 to a constant and lets do the same to *

File: components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java (hunk 1 above)
Comment (janakamarasena) [reply components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java:149]: lets write a unit test for this

File: components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java [hunk 3]
```diff
@@ -135,6 +136,50 @@ public boolean doPostDeleteApplication(String applicationName, String tenantDoma
...
+     * @param userName        Username of the person who does the deletion.
+     * @return true.
+     * @throws IdentityApplicationManagementException IdentityApplicationManagementException.
+     */
+    @Override
+    public boolean doPostUpdateApplication(ServiceProvider serviceProvider, String tenantDomain, String userName)
+            throws IdentityApplicationManagementException {
+
+        ConsentManager consentManager = IdentityConsentDataHolder.getInstance().getConsentManager();
+        if (serviceProvider.isApplicationEnabled()) {
```
Comment (janakamarasena) [inline components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java:153]: move this code block up. we don't need to do `ConsentManager consentManager = IdentityConsentDataHolder.getInstance().getConsentManager();` if the app is not disabled.

File: components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java (hunk 1 above)
Comment (VivekVinushanth) [reply components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/main/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListener.java:132]: https://github.com/wso2/carbon-identity-framework/pull/5732/commits/4f0851d547ebf25fe2eb6f4867950a1153a67222

File: components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/test/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListenerTest.java [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+import static org.wso2.carbon.base.MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+import static org.wso2.carbon.base.MultitenantConstants.SUPER_TENANT_ID;
+import static org.wso2.carbon.identity.application.authentication.framework.handler.request.impl.consent.constant.SSOConsentConstants.CONFIG_ELEM_CONSENT;
+import static org.wso2.carbon.identity.consent.mgt.listener.ConsentDeletionAppMgtListener.CONSENT_SEARCH_PII_PRINCIPAL_ID;
+
+
+public class ConsentDeletionAppMgtListenerTest {
```
Comment (janakamarasena) [inline components/consent-mgt/org.wso2.carbon.identity.consent.mgt/src/test/java/org/wso2/carbon/identity/consent/mgt/listener/ConsentDeletionAppMgtListenerTest.java:87]: Lets add a class comment here

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**